server/neptune/workflows/internal/pr/runner.go

```diff
@@ -0,0 +1,217 @@
+		RevisionReceiver:      &revisionReceiver,
+		ShutdownSignalChannel: workflow.GetSignalChannel(ctx, receiver.ShutdownSignalID),
+		ShutdownReceiver:      &shutdownReceiver,
+		Scope:                 scope,
+		TFWorkflow:            tfWorkflow,
+		TFStateReceiver:       StateReceiver{InternalNotifiers: internalNotifiers},
+	}
+}
+
+// Run handles managing the workflow's context lifecycles as new signals/poll events are received and
+// change the current PRAction status
+func (r *Runner) Run(ctx workflow.Context) error {
+	var action Action
+	var prRevision receiver.Revision
+
+	//TODO: add approve signal, timeouts, poll variation of shutdown signal
+	s := workflow.NewSelector(ctx)
+	s.AddReceive(r.RevisionSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		prRevision = r.RevisionReceiver.Receive(c, more)
+		action = onNewRevision
+	})
+
+	s.AddReceive(r.ShutdownSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		r.ShutdownReceiver.Receive(c, more)
+		action = onShutdown
+		r.state = complete
+	})
+
+	_, cancel := workflow.WithCancel(ctx)
+	for {
+		s.Select(ctx)
+		switch action {
+		case onNewRevision:
+			revisionCtx := workflow.WithValue(ctx, internalContext.SHAKey, prRevision.Revision)
+			workflow.GetLogger(revisionCtx).Info("received revision")
+			if process := r.shouldProcessRevision(prRevision); !process {
+				continue
+			}
+			// cancel in progress workflow (if it exists) and start up a new one
+			cancel()
+			revisionCtx, cancel = workflow.WithCancel(revisionCtx)
+			r.state = working
+			r.lastAttemptedRevision = prRevision.Revision
+			workflow.Go(revisionCtx, func(c workflow.Context) {
+				r.processRevision(c, prRevision)
+			})
+		case onShutdown:
+			//todo: maybe optimize by cancelling in progress child workflows
+			workflow.GetLogger(ctx).Info("received shutdown request")
+			return nil
+		}
+	}
+}
+
+func (r *Runner) shouldProcessRevision(prRevision receiver.Revision) bool {
+	// ignore reruns when revision is still in progress
+	if r.lastAttemptedRevision == prRevision.Revision && r.state != waiting {
+		return false
+	}
+	return true
+}
+
+// processRevision handles spinning off child Terraform workflows per root and
+// dealing with any failed policies by reviewing set of approvals
+func (r *Runner) processRevision(ctx workflow.Context, prRevision receiver.Revision) {
+	defer func() {
+		r.state = waiting
+	}()
+	failedPolicies := make(map[string]activities.PolicySet)
+	var futures []workflow.ChildWorkflowFuture
+	var rootInfos []RootInfo
+	for _, root := range prRevision.Roots {
+		ctx = workflow.WithValue(ctx, internalContext.ProjectKey, root.Name)
+		future, rootInfo, err := r.processRoot(ctx, root, prRevision)
+		if err != nil {
+			continue
+		}
+		futures = append(futures, future)
+		rootInfos = append(rootInfos, rootInfo)
+	}
+	for i, future := range futures {
+		failedRootPolicies, err := r.awaitWorkflow(ctx, future, rootInfos[i])
```

Comment:
I dont think this will work, because you're attaching the notifiers after the workflows are already in progress so you'll miss events/won't notify in real time.  I think the notifiers have to be executed with the workflows as well.

Reply:
Discussed and am going to start out with using one selector to listen to all child workflow futures/state change events and test this out.